src/WidgetUtils.js

```diff
@@ -136,4 +136,56 @@ export default class WidgetUtils {
             MatrixClientPeg.get().on('accountData', onAccountData);
         });
     }
+
+    /**
+     * Returns a promise that resolves when a widget with the given
+     * ID has been added as a room widget in the given room (ie. the
+     * room state event arrives) or rejects after a timeout
+     *
+     * @param {string} widgetId The ID of the widget to wait for
+     * @param {string} roomId The ID of the room to wait for the widget in
+     * @param {boolean} add True to wait for the widget to be added,
+     *     false to wait for it to be deleted.
+     * @returns {Promise} that resolves when the widget is available
+     */
+    static waitForRoomWidget(widgetId, roomId, add) {
```

Comment:
Any chance the shared logic between this and `waitForUserWidget` could be factored out? They both do very similar things on slightly different inputs..

Reply:
Yeah, I considered this but I think all of the bits of these functions are subtly different, plus there's already the fact that both of them wait for either addition or deletion. I think in this case having two less complex functions might be better than having one more complex one.

Reply:
I'm not necessarily suggesting one function; but I notice that
 - `eventsInIntendedState` is very similar to eventInIntendedState`
 - they both add a listener to matrix client
 - they both set a timeout
   - the timeout is 10 seconds with both
   - the timeout is cleared when the event is received by the matrix client
 
Actually, I noticed they've already diverged slightly: 
https://github.com/matrix-org/matrix-react-sdk/pull/1958/files/36574ca0fb1b0a14a5886a953b3d6dbaad2ea7d2..3e6b3215cfe15532e19fd48288caf9fe23fd465e#diff-18c7dfd796a5f668feeee73e460716f3R125

gets the event from the account data again instead of reusing `startingAccountDataEvent`, whereas `waitForRoomWidget` reuses `startingWidgetEvents`.

Reply:
Not sure what you mean - they both re-fetch the current state: https://github.com/matrix-org/matrix-react-sdk/blob/3e6b3215cfe15532e19fd48288caf9fe23fd465e/src/WidgetUtils.js#L119 and https://github.com/matrix-org/matrix-react-sdk/blob/3e6b3215cfe15532e19fd48288caf9fe23fd465e/src/WidgetUtils.js#L178

Reply:
Oh. I'm a crank apparently. 

IRL we discussed and concluded that these two things are indeed subtly different, pretty much by consequence of the underlying API being used subtly differently depending on whether the widget is in account data or whether it is in a room.